Allow overriding settings in project settings

bordaigorl opened the pull request:
It may be that in certain projects you want to use a different Opam switch, which would require one to change `coq_install_dir`.
With this change, if project-specific settings are found, they override the global ones.
Maybe we should consider prefixing the keys with `coq.`.

bordaigorl committed:
Allow overriding settings in project settings

Calvin-L commented on CoqPlugin.py:
```diff
@@ -763,6 +763,9 @@ def run(self, edit):
             worker = None
 
         settings = sublime.load_settings("CoqInteractive.sublime-settings")
+        for k in ("coq_install_dir", "view_style", "move_cursor_after_command", "show_goals"):
+            if self.view.settings().has("coq."+k):
+                settings.set(k, self.view.settings()["coq."+k])
```
Sublime's `Settings` objects automagically pick up changes when the user modifies+saves a settings file. (This [seems to come up a lot](https://github.com/Calvin-L/sublime-coq-plugin/pull/2#issuecomment-973656888).) Do you know how `set` interacts with that behavior?

I don't know the answer myself, but I suspect that any `set` properties will be clobbered when the settings change. If so, this commit introduces weird behaviors like:
 1. Make a project with a setting to override `coq.show_goals`.
 2. Launch the plugin in that project (which runs this code)
 3. Interact with Coq for a while.
 4. Modify your global settings.
 5. Sublime reloads the `Settings` object, clobbering the project-specific setting set in step (2)
 6. Interact with Coq again ==> Surprise! It now ignores the project-specific setting for `coq.show_goals` until you kill and re-launch Coq.

Again, it's not clear to me, but I suspect that this happens with _any_ global settings change, not just changes to the plugin-specific settings.

----

All that said, project-specific overrides are a great idea and I would probably use them myself. For now, let's just limit the overrideable properties to the "per-session-constant" ones `("coq_install_dir", "view_style")`. I'll think about what to do for the others when I get a chance later this week.

Calvin-L committed:
Resolve settings using view-speific settings first

This fixes a number of weird behaviors and latent bugs surrounding use
of `Settings.set(...)`.  There doesn't seem to be a way to make such
an impure approach ironclad; it is better to adjust how settings are
read than to modify the in-memory settings object.

This commit fixes the following strange/buggy behaviors:

  1. When a project-specific setting is removed from the project
     definition, the plugin would not go back to using the global
     setting.
  2. Project-specific overrides for dynamic settings like "show_goals"
     would be ignored until restarting the Coq session.
  3. Project-specific overrides for dynamic settings like "show_goals"
     could be forgotten if the global settings were modified while
     Coq was running.

(There's a lesson in here somewhere about the benefits of using a more
complex read algorithm rather than trying to keep the system state as
one where the shortest possible read algorithm behaves correctly.
Modifying the in-memory settings to make `Settings.get()` return the
correct value is a lot more error-prone than modifying the read
algorithm.)

This commit also adds text to README.md about the new project-specific
settings feature.

Calvin-L commented:
To handle the weirdness around settings reload, I've changed how this PR works. Instead of modifying the in-memory settings when Coq starts, the (new) procedure for reading settings will first consult the view-specific settings and then the plugin-specific settings.

Thanks a ton for this change; I expect to use this feature myself. :)
